xbmc/windowing/wayland/WinSystemWayland.cpp

```diff
@@ -289,46 +364,62 @@ bool CWinSystemWayland::SetFullScreen(bool fullScreen, RESOLUTION_INFO& res, boo
-    m_shellSurface.set_toplevel();
+    m_shellSurface->SetWindowed();
   }
 
   return true;
 }
 
-void CWinSystemWayland::HandleSurfaceConfigure(wayland::shell_surface_resize edges, std::int32_t width, std::int32_t height)
+void CWinSystemWayland::HandleSurfaceConfigure(std::int32_t width, std::int32_t height)
 {
+  // Wayland will tell us here the size of the surface that was actually created,
+  // which might be different from what we expected e.g. when fullscreening
+  // on an output we chose - the compositor might have decided to use a different
+  // output for example
+  // It is very important that the EGL native module and the rendering system use the
+  // Wayland-announced size for rendering or corrupted graphics output will result.
+  
   CLog::Log(LOGINFO, "Got Wayland surface size %dx%d", width, height);
+  {
+    CSingleLock lock(m_configurationMutex);
 
-  // Mark everything opaque so the compositor can render it faster
-  wayland::region_t opaqueRegion = m_connection->GetCompositor().create_region();
-  opaqueRegion.add(0, 0, width, height);
-  m_surface.set_opaque_region(opaqueRegion);
-  // No surface commit, EGL context will do that when it changes the buffer
+    // Mark everything opaque so the compositor can render it faster
+    wayland::region_t opaqueRegion = m_connection->GetCompositor().create_region();
+    opaqueRegion.add(0, 0, width, height);
+    m_surface.set_opaque_region(opaqueRegion);
+    // No surface commit, EGL context will do that when it changes the buffer
 
-  if (m_nWidth == width && m_nHeight == height)
-  {
-    // Nothing to do
-    return;
-  }
+    if (m_nWidth == width && m_nHeight == height)
+    {
+      // Nothing to do
+      return;
+    }
 
-  m_nWidth = width;
-  m_nHeight = height;
-  // Update desktop resolution
-  auto& res = CDisplaySettings::GetInstance().GetCurrentResolutionInfo();
-  res.iWidth = width;
-  res.iHeight = height;
-  res.iScreenWidth = width;
-  res.iScreenHeight = height;
-  res.iSubtitles = (int) (0.965 * height);
-  g_graphicsContext.ResetOverscan(res);
-  CDisplaySettings::GetInstance().ApplyCalibrations();
+    m_nWidth = width;
+    m_nHeight = height;
+    // Update desktop resolution
+    auto& res = CDisplaySettings::GetInstance().GetCurrentResolutionInfo();
+    res.iWidth = width;
+    res.iHeight = height;
+    res.iScreenWidth = width;
+    res.iScreenHeight = height;
+    res.iSubtitles = (int) (0.965 * height);
+    g_graphicsContext.ResetOverscan(res);
+    CDisplaySettings::GetInstance().ApplyCalibrations();
+  }
+  
   // Force resolution update
+  // SetVideoResolution() automatically delegates to main thread via internal
+  // message if called from other threads
+  // This will call SetFullScreen() with the new resolution, which also updates
+  // the size of the egl_window etc.
+  // The call always blocks, so the configuration lock must be released beforehand.
```

Comment:
I also don't like it, but can't help it for now. I had to rework lots of this stuff for #8, TODO is included there.